File: app/src/main/java/co/k2/newsbits/data/source/local/LocalDataSourceImpl.java
package co.k2.newsbits.data.source.local;

import java.util.List;

import javax.inject.Inject;

import co.k2.newsbits.data.models.Article;
import io.reactivex.Completable;
import io.reactivex.Single;

/**
 * Copyright (C) 2019 K2 CODEWORKS
 * All rights reserved
 *
 * @author Kislay
 * @since 28/03/19
 */

public class LocalDataSourceImpl implements LocalDataSource {

    private final ArticlesDB articlesDB;

    @Inject
    public LocalDataSourceImpl(ArticlesDB db) {
        this.articlesDB = db;
    }

    @Override
    public Single<List<Article>> getHeadlines(String country) {
        return articlesDB.savedBeaconDao().getAll();
    }

    @Override
    public Completable saveToCache(List<Article> articles) {
        return articlesDB.savedBeaconDao().insertAll(articles);
    }

    @Override
    public Completable clearCache() {
        return articlesDB.savedBeaconDao().deleteAll();
    }

}

File: app/src/main/java/co/k2/newsbits/data/source/local/LocalDataSourceImpl.kt
package co.k2.newsbits.data.source.local

import co.k2.newsbits.data.models.Article
import co.k2.newsbits.data.models.ArticleCacheMetaData
import javax.inject.Inject

class LocalDataSourceImpl @Inject constructor(
    private val articlesDB: ArticlesDB
) : LocalDataSource {

    override suspend fun clearCache() {
        articlesDB.savedArticlesDao().deleteAll()
    }

    override suspend fun saveToCache(articles: List<Article>) {
        articlesDB.savedArticlesDao().insertAll(articles)
    }

    override suspend fun getHeadlines(country: String?): List<Article> {
        return articlesDB.savedArticlesDao().all ?: emptyList()
    }

    override suspend fun getCacheMetaData(): ArticleCacheMetaData? {
        return articlesDB.articlesMetaDataDao().getMetadata
    }

    override suspend fun saveCacheMetaData(articleCacheMetaData: ArticleCacheMetaData) {
        articlesDB.articlesMetaDataDao().insert(articleCacheMetaData)
    }
}
